### PR TITLE
Add multi email support

### PR DESCRIPTION
**Pull request by lgarvey**
WARNING: This PR contains a data migration that copies the `User.email` field into the `User.emails` list

This PR adds multiple emails per user support. This is achieved by adding an `EmailAddress` model that FK's to the `User` model.

The `User.email` field has been retained as removing it causes issues with auth model checks, particularly the requirement that there is a field on the User model with `unique=True`

Local authentication, e.g. logging into the django admin uses the `User.email` field. Whereas saml and email token authentication predominantly uses the `User.emails` list.

`User.objects.get_or_create()` and `User.save()` have been overridden to make sure that the `User.emails` list is queried and that the `User.email` field is automatically copied into the `User.emails` list. This is to leverage the `unique=True` constraint on the `EmailAddress.email` field and avoid having to manage cross model/field unique constraints at the code level.

The `oauth2.Application` model now has an `email_ordering` field to determine the precedence of emails that are supplied to the consuming application.  The first email to match that list will be supplied to the application on the user profile page under the `email` key. All other emails are listed on the user profile page under the `related_emails` key.

### CODE REVIEW

File: config/settings.py [hunk 1]
```diff
@@ -309,3 +309,6 @@
 
 # google analytics
 GOOGLE_ANALYTICS_CODE = env('GOOGLE_ANALYTICS_CODE', default=None)
+
+# the default order applied to user emails
+DEFAULT_EMAIL_ORDER = env('EMAIL_PRORITY_LIST', default=[])
```
Comment (r4vi) [inline config/settings.py]: can you make it clear what is valid to set in this env var?

Comment (lgarvey) [reply config/settings.py]: Yep, I'll add a better comment

File: config/urls.py [hunk 2]
```diff
@@ -4,11 +4,13 @@
 
 from sso.healthcheck.views import HealthCheckView
 from sso.oauth2.views import CustomAuthorizationView
+from sso.samlauth.views import session_logout
 from . import api_urls
 
 urlpatterns = [
     url(r'^admin/', admin.site.urls),
     url(r'^saml2/', include('sso.samlauth.urls')),
+    url(r'^logout/$', session_logout, name='session-logout'),
```
Comment (marcofucci) [inline config/urls.py]: this is OK but I tend not to put views directly here. As `session_logout` comes from `sso.samlauth` I would generally refactor
`url(r'^saml2/', include('sso.samlauth.urls')),`
into something like:
`url(r'^', include('sso.samlauth.urls')),`
and add the `saml2/` prefix in the `'sso.samlauth.urls'` if needed.

Comment (marcofucci) [reply config/urls.py]: I think there's a bit of confusion around saml2 login/logout urls vs generic ones.
We started with saml2 only but now we have different options the urls might need to be revisited a bit.

**Comment by lgarvey** [comment]
This is going to get removed. Unfortunately I'm testing some stuff with
Core with this PR.

**Comment by lgarvey** [comment]
Yep, makes sense.  Mat specifically wanted /logout/.  I'll remove the
/saml2/ prefix from the main include and add individually in the
saml2auth/urls.py file, then move the login url back into the module.

File: config/settings.py [hunk 3]
```diff
@@ -187,7 +187,8 @@
                     (SAML_ACS_URL, saml2.BINDING_HTTP_POST),
                 ],
                 'single_logout_service': [
-                    (SAML_REDIRECT_RETURN_HOST + '/saml2/ls/post/', saml2.BINDING_HTTP_POST),
+                    'https://adfsmobile.azurecore.com/adfs/ls/?wa=wsignout1.0'
+                    #(SAML_REDIRECT_RETURN_HOST + '/saml2/ls/post/', saml2.BINDING_HTTP_POST),
```
[the remark this answers is not in the document]
Comment (lgarvey) [reply config/settings.py]: This will be gone in a sec. I hijacked this PR to test some core stuff.

**Comment by lgarvey** [comment]
I've moved the logout view into the localauth app, which isn't using a url
prefix:     url(r'^', include('sso.localauth.urls', namespace='localauth')),

Kind of makes more sense there anyway.

**Pull request merged**